File: core/src/com/jupiter/europa/screen/overlay/PartyOverlay.java
package com.jupiter.europa.screen.overlay;

import com.badlogic.ashley.core.Entity;
import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.graphics.g2d.TextureAtlas;
import com.badlogic.gdx.scenes.scene2d.Actor;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.ui.Image;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import com.badlogic.gdx.scenes.scene2d.utils.SpriteDrawable;
import com.badlogic.gdx.utils.Disposable;
import com.jupiter.europa.EuropaGame;
import com.jupiter.europa.entity.EntityEventArgs;
import com.jupiter.europa.entity.Families;
import com.jupiter.europa.entity.Mappers;
import com.jupiter.europa.entity.Party;
import com.jupiter.europa.entity.component.MovementResourceComponent;
import com.jupiter.europa.entity.component.ResourceComponent;
import com.jupiter.europa.entity.stats.characterclass.CharacterClass;
import com.jupiter.europa.entity.stats.race.Race;
import com.jupiter.europa.io.FileLocations;
import com.jupiter.europa.scene2d.ui.ResourceBar;
import com.jupiter.europa.screen.OverlayableScreen;
import com.jupiter.ganymede.event.Event;
import com.jupiter.ganymede.event.Listener;

import java.util.Arrays;

/**
 * Created by nathan on 5/18/15.
 */
public class PartyOverlay extends Scene2DOverlay implements Disposable {

    // Constants
    public static final float PARTY_MEMBER_SCALE = 3.0f;


    // Fields
    private final Event<EntityEventArgs> entityClicked = new Event<>();
    private final Table table = new Table();
    private final Table anchor = new Table();
    private final Party party;


    // Initialization
    public PartyOverlay(Party party) {
        super(false);
        this.party = party;
        this.initComponents();
    }


    // Public Methods
    public boolean addEntityClickListener(Listener<EntityEventArgs> listener) {
        return this.entityClicked.addListener(listener);
    }

    public boolean removeEntityClickListener(Listener<EntityEventArgs> listener) {
        return this.entityClicked.removeListener(listener);
    }

    @Override
    public void added(OverlayableScreen screen) {
        super.added(screen);
        this.anchor.setFillParent(true);
        this.anchor.add(this.table).expand().top().right().pad(10);
        this.getStage().addActor(anchor);

        screen.addTintChangedListener(args -> this.setTint(args.newValue));
    }

    @Override
    public void dispose() {
        for (Actor act : this.table.getChildren()) {
            if (act instanceof Disposable) {
                ((Disposable) act).dispose();
            }
        }
    }


    // Private Methods
    private void initComponents() {
        Arrays.asList(this.party.getActivePartyMembers()).stream()
                .forEach(entity -> {
                    EntityActor act = new EntityActor(entity);
                    act.addListener(new ClickListener() {
                        @Override
                        public void clicked(InputEvent e, float x, float y) {
                            PartyOverlay.this.onEntityClick(entity);
                        }
                    });
                    this.table.add(act).center();
                });
    }

    private void onEntityClick(Entity entity) {
        this.entityClicked.dispatch(new EntityEventArgs(entity));
    }


    // Inner Classes
    private static class EntityActor extends Table implements Disposable {

        // Constants
        private static final int BAR_HEIGHT = 3;
        private static final int BAR_PADDING = 1;


        // Static Methods
        private static Sprite getSprite(Entity entity) {
            if (Families.classed.matches(entity) && Families.raced.matches(entity)) {
                CharacterClass characterClass = Mappers.characterClass.get(entity).getCharacterClass();
                Race race = Mappers.race.get(entity).getRace();
                String textureClassString = characterClass.getTextureSetName();
                String texture = race.getTextureString() + "-" + textureClassString + "-" + MovementResourceComponent.FRONT_STAND_TEXTURE_NAME;
                return new Sprite(EuropaGame.game.getAssetManager().get(FileLocations.SPRITES_DIRECTORY.resolve(
                        "CharacterSprites.atlas").toString(), TextureAtlas.class).findRegion(texture));
            }
            return new Sprite();
        }


        // Fields
        private final Image entityPreview = new Image();
        private final Entity entity;


        // Initialization
        private EntityActor(Entity entity) {
            this.entity = entity;
            this.initComponents();
        }


        // Public Methods
        @Override
        public void dispose() {
            for (Actor act : this.getChildren()) {
                if (act instanceof Disposable) {
                    ((Disposable) act).dispose();
                }
            }
        }


        // Private Methods
        private void initComponents() {
            Sprite sprite = getSprite(this.entity);
            this.entityPreview.setDrawable(new SpriteDrawable(sprite));

            final float width = sprite.getWidth() * PARTY_MEMBER_SCALE;

            this.add(this.entityPreview).center().width(width).height(sprite.getHeight() * PARTY_MEMBER_SCALE);
            this.row();

            for (ResourceComponent.Resources resource : ResourceComponent.Resources.values()) {
                this.add(new ResourceBar(this.entity, resource)).center().expandX().fillX().height(BAR_HEIGHT).padTop(BAR_PADDING);
                this.row();
            }
        }
    }
}

File: core/src/com/jupiter/europa/screen/overlay/PartyOverlay.kt
/*
 * The MIT License
 *
 * Copyright 2015 Nathan Templon.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

package com.jupiter.europa.screen.overlay

import com.badlogic.ashley.core.Entity
import com.badlogic.gdx.graphics.g2d.Sprite
import com.badlogic.gdx.graphics.g2d.TextureAtlas
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.ui.Image
import com.badlogic.gdx.scenes.scene2d.ui.Table
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener
import com.badlogic.gdx.scenes.scene2d.utils.SpriteDrawable
import com.badlogic.gdx.utils.Disposable
import com.jupiter.europa.EuropaGame
import com.jupiter.europa.entity.EntityEventArgs
import com.jupiter.europa.entity.Families
import com.jupiter.europa.entity.Mappers
import com.jupiter.europa.entity.Party
import com.jupiter.europa.entity.component.MovementResourceComponent
import com.jupiter.europa.entity.component.ResourceComponent
import com.jupiter.europa.io.FileLocations
import com.jupiter.europa.scene2d.ui.ResourceBar
import com.jupiter.europa.screen.OverlayableScreen
import com.jupiter.ganymede.event.Event
import com.jupiter.ganymede.event.Listener

/**
 * Created by nathan on 5/18/15.
 */
public class PartyOverlay(private val party: Party) : Scene2DOverlay(false), Disposable {


    // Fields
    private val entityClicked = Event<EntityEventArgs>()
    private val table = Table()
    private val anchor = Table()


    init {
        this.initComponents()
    }


    // Public Methods
    public fun addEntityClickListener(listener: Listener<EntityEventArgs>): Boolean {
        return this.entityClicked.addListener(listener)
    }

    public fun removeEntityClickListener(listener: Listener<EntityEventArgs>): Boolean {
        return this.entityClicked.removeListener(listener)
    }

    override fun added(screen: OverlayableScreen) {
        super<Scene2DOverlay>.added(screen)
        this.anchor.setFillParent(true)
        this.anchor.add(this.table).expand().top().right().pad(10f)
        this.stage?.addActor(anchor)

        screen.addTintChangedListener(Listener { args -> this.tint = args.newValue })
    }

    override fun dispose() {
        for (act in this.table.getChildren()) {
            if (act is Disposable) {
                act.dispose()
            }
        }
    }


    // Private Methods
    private fun initComponents() {
        this.party.getActivePartyMembers().forEach { entity ->
            val actor = EntityActor(entity)
            actor.addListener(object : ClickListener() {
                override fun clicked(e: InputEvent, x: Float, y: Float) {
                    this@PartyOverlay.onEntityClick(entity)
                }
            })
            this.table.add(actor).center()
        }
    }

    private fun onEntityClick(entity: Entity) {
        this.entityClicked.dispatch(EntityEventArgs(entity))
    }


    // Inner Classes
    public class EntityActor(private val entity: Entity) : Table(), Disposable {


        // Fields
        private val entityPreview = Image()


        init {
            this.initComponents()
        }


        // Public Methods
        override fun dispose() {
            for (act in this.getChildren()) {
                if (act is Disposable) {
                    act.dispose()
                }
            }
        }


        // Private Methods
        private fun initComponents() {
            val sprite = getSprite(this.entity)
            this.entityPreview.setDrawable(SpriteDrawable(sprite))

            val width = sprite.getWidth() * PARTY_MEMBER_SCALE

            this.add(this.entityPreview).center().width(width).height(sprite.getHeight() * PARTY_MEMBER_SCALE)
            this.row()

            for (resource in ResourceComponent.Resources.values()) {
                this.add(ResourceBar(this.entity, resource)).center().expandX().fillX().height(BAR_HEIGHT.toFloat()).padTop(BAR_PADDING.toFloat())
                this.row()
            }
        }

        companion object {

            // Constants
            private val BAR_HEIGHT = 3
            private val BAR_PADDING = 1


            // Static Methods
            private fun getSprite(entity: Entity): Sprite {
                if (Families.classed.matches(entity) && Families.raced.matches(entity)) {
                    val characterClass = Mappers.characterClass.get(entity).characterClass
                    val race = Mappers.race.get(entity).race
                    val textureClassString = characterClass.textureSetName
                    val texture = (race?.textureString ?: "") + "-" + textureClassString + "-" + MovementResourceComponent.FRONT_STAND_TEXTURE_NAME
                    return Sprite(EuropaGame.game.assetManager!!.get(FileLocations.SPRITES_DIRECTORY.resolve("CharacterSprites.atlas").toString(), javaClass<TextureAtlas>()).findRegion(texture))
                }
                return Sprite()
            }
        }
    }

    companion object {

        // Constants
        public val PARTY_MEMBER_SCALE: Float = 3.0f
    }
}
